modules/images/image-loading-optimization/optimization.php

```diff
@@ -30,26 +30,27 @@ function ilo_maybe_add_template_output_buffer_filter() {
  * @since n.e.x.t
  * @access private
  *
- * @param array $lcp_images_by_minimum_viewport_widths LCP images keyed by minimum viewport width, amended with attributes key for the IMG attributes.
+ * @param array<int, array{attributes: array{src?: string, srcset?: string, sizes?: string, crossorigin?: string}}|false> $lcp_elements_by_minimum_viewport_widths LCP images keyed by minimum viewport width, amended with attributes key for the IMG attributes.
```

Comment:
> Maybe a personal preference, but I find this notation overkill. This isn't used anywhere in WordPress. If we want to describe more complex shapes like this, we should probably introduce a class for them. This would be safer in any case than using an associative array which by definition takes any arbitrary data.

Actually, it is used in WordPress, specifically [in the `sodium_compat` code](https://github.com/search?q=repo%3AWordPress%2Fwordpress-develop%20%2Farray%7B%2F&type=code). Granted, this is not really organic WordPress code. In any case, with precise phpdoc typing here, PHPStan will detect any anomalies with static analysis. So it's trying to gain the same benefits as TypeScript gives.

> I think a simple `ILO_LCP_Element` class or similar would be helpful here. Then this would become `ILO_LCP_Element[]`.

Yeah, I like that idea. However, believe it or not, I was wanting to limit the refactor of the existing code in this PR once the test coverage was completed. With the coverage in place, refactoring will be safer as I'll be able to detect regressions better.

> Separate minor question: Why was this renamed to "LCP elements" instead of "LCP images"? Would it ever not be an LCP image? If so, the choice of fields would be a bit odd, since almost all of them are specific to images.

Yes, for example in https://github.com/WordPress/performance/pull/914 this is extended to support elements with CSS background images.

Reply:
> Yeah, I like that idea. However, believe it or not, I was wanting to limit the refactor of the existing code in this PR once the test coverage was completed. With the coverage in place, refactoring will be safer as I'll be able to detect regressions better.

+1, that makes total sense to me. Maybe open an issue already?

Reply:
I've just now added it as a todo item on the [overview issue](https://github.com/WordPress/performance/issues/869).